builtin/logical/transit/lock_manager.go

```diff
@@ -0,0 +1,337 @@
+package transit
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/vault/logical"
+)
+
+const (
+	shared    = false
+	exclusive = true
+)
+
+type lockManager struct {
+	// A lock for each named key
+	locks map[string]*sync.RWMutex
+
+	// A mutex for the map itself
+	lockMutex sync.RWMutex
+
+	// If caching is enabled, the map of name to in-memory policy cache
+	cache map[string]*Policy
+
+	// Used for global locking, and as the cache map mutex
+	globalMutex sync.RWMutex
```

Comment:
It's to reduce lock contention. `globalMutex` is used in various places not related to a particular lock. For instance, `LockAll`/`UnlockAll` -- other goroutines could still perform lookups of their policy locks while the goroutine that locked the global lock is executing. This is super critical-path stuff, so I tried to keep as little lock contention as possible